### PR TITLE
Add `secrets` endpoint

### PR DESCRIPTION
**Pull request by shahnami**
Adds a `createSecrets` function as part of `defender-autotask-client`

### CODE REVIEW

File: packages/base/src/api/client.ts [hunk 1]
```diff
@@ -18,24 +18,24 @@ export abstract class BaseApiClient {
     this.apiSecret = params.apiSecret;
   }
 
-  protected async init(): Promise<AxiosInstance> {
```
Comment (shahnami) [inline packages/base/src/api/client.ts:21]: Doing this to avoid having to create an entire new `defender-secrets-client` for one endpoint.

**Review by spalladino — CHANGES_REQUESTED** [review]
I believe there's an issue with the `overrideApiUrl`. The `api` object in each client is initialized once and reused for all subsequent requests, so if we change the apiUrl in one of the methods of the client, the `api` will end up initialized with the url of that method. This means that if you first call `createSecrets` but then try calling any other method in the client (or vice versa) the request will fail, since the base url will be incorrect.

My suggestion is to use the `DEFENDER_AUTOTASK_API_URL` as the base URL for the secrets endpoint as well, which could look like `https://defender-api.openzeppelin.com/autotask/secrets`. Note that today we already have a duplication in the path: autotasks are listed by querying `https://defender-api.openzeppelin.com/autotask/autotasks`.

**Comment by shahnami** [comment]
> I believe there's an issue with the `overrideApiUrl`. The `api` object in each client is initialized once and reused for all subsequent requests, so if we change the apiUrl in one of the methods of the client, the `api` will end up initialized with the url of that method. This means that if you first call `createSecrets` but then try calling any other method in the client (or vice versa) the request will fail, since the base url will be incorrect.

Just tested, and works as expected. This is because we don't call `init` once per client initialisation. We call it every time `apiCall` is called, which means it gets re-initialised for every function.

> My suggestion is to use the `DEFENDER_AUTOTASK_API_URL` as the base URL for the secrets endpoint as well, which could look like `https://defender-api.openzeppelin.com/autotask/secrets`. Note that today we already have a duplication in the path: autotasks are listed by querying `https://defender-api.openzeppelin.com/autotask/autotasks`.

However, I do agree that this might be a better approach.

**Comment by spalladino** [comment]
> This is because we don't call init once per client initialisation. We call it every time apiCall is called, which means it gets re-initialised for every function.

Hmm but `init` checks if `api` has been created or not, and if it was already set, then it skips re-initialization

https://github.com/OpenZeppelin/defender-client/blob/ab54f4cf0f614d9e80cc827ae7e94d44398b9a0b/packages/base/src/api/client.ts#L21-L28

**Comment by shahnami** [comment]
> > This is because we don't call init once per client initialisation. We call it every time apiCall is called, which means it gets re-initialised for every function.
> 
> Hmm but `init` checks if `api` has been created or not, and if it was already set, then it skips re-initialization
> 
> https://github.com/OpenZeppelin/defender-client/blob/ab54f4cf0f614d9e80cc827ae7e94d44398b9a0b/packages/base/src/api/client.ts#L21-L28

Yeah that's odd. I tested in the create-autotask example package with this:

```
  const createSecrets = await client.createSecrets({ deletes: [], secrets: { "foo": "bar" } });
  const createdAutotask = await client.create(myAutotask);
  console.log(createSecrets);
  console.log(createdAutotask)
```
**Output:**
```
{ secretNames: [ 'hello', 'foo' ] }
{
  autotaskId: 'f4b9d056-41ba-424a-abbb-88d578dfd1be',
  name: 'my-autotask',
  paused: false,
  trigger: { type: 'schedule', frequencyMinutes: 1500 },
  encodedZippedCode: 'UEsDBAoAAAAAAAAAQVDA91VHFwAAABcAAAAHAAAAZGVwcy5qc2V4cG9ydHMuZm9vID0gKCkgPT4gNDI7UEsDBAoAAAAAAAAAQVCEyiyCiQAAAIkAAAAIAAAAaW5kZXguanNjb25zdCBkZXBzID0gcmVxdWlyZSgnLi9kZXBzJyk7CgpleHBvcnRzLmhhbmRsZXIgPSBhc3luYyBmdW5jdGlvbigpIHsKICBjb25zdCB2YWx1ZSA9IGRlcHMuZm9vKCk7CiAgY29uc29sZS5sb2codmFsdWUpOwogIHJldHVybiB2YWx1ZTsKfVBLAQIUAAoAAAAAAAAAQVDA91VHFwAAABcAAAAHAAAAAAAAAAAAAAAAAAAAAABkZXBzLmpzUEsBAhQACgAAAAAAAABBUITKLIKJAAAAiQAAAAgAAAAAAAAAAAAAAAAAPAAAAGluZGV4LmpzUEsFBgAAAAACAAIAawAAAOsAAAAAAA=='
}
```

File: packages/autotask-client/src/api.ts [hunk 2]
```diff
@@ -108,4 +113,10 @@ export class AutotaskClient extends BaseApiClient {
       return await api.put(`/autotasks/${autotaskId}/code`, { encodedZippedCode });
     });
   }
+
+  public async createSecrets(data: SaveSecretsRequest): Promise<GetSecretsResponse> {
+    return this.apiCall(async (api) => {
+      return await api.post(`/autotasks/secrets`, data);
```
Comment (spalladino) [inline packages/autotask-client/src/api.ts:119]:
```suggestion
      return await api.post(`/secrets`, data);
```

You shouldn't need the leading `/autotasks` here, that's provided in the base `DEFENDER_AUTOTASK_API_URL`

**Comment by spalladino** [comment]
> Yeah that's odd. I tested in the create-autotask example package with this:

Agree, that's definitely odd

File: packages/autotask-client/src/api.ts (hunk 2 above)
Comment (shahnami) [reply packages/autotask-client/src/api.ts]: Yep, API gateway mapping confused me there for a sec.

**Review by spalladino — APPROVED** [review]

File: packages/autotask-client/src/models/autotask.ts [hunk 3]
```diff
@@ -33,3 +33,15 @@ export interface Autotask extends Pick<CreateAutotaskRequest, 'name' | 'relayerI
   createdAt?: string;
   codeDigest?: string;
 }
+
+export interface SaveSecretsRequest {
+  deletes: string[];
```
Comment (BjMrq) [inline packages/autotask-client/src/models/autotask.ts:38]: should the `delete` be optional?

Comment (mok0230) [reply packages/autotask-client/src/models/autotask.ts:38]: @shahnami note sure if you saw this one as it was after the PR was merged, but it does seem like these should be optional params I think

Comment (shahnami) [reply packages/autotask-client/src/models/autotask.ts:38]: Missed this one, thanks for bringing it to my attention :)

This type was copied from our shared models. Perhaps we _should_ make it optional in Defender too.

**Pull request merged**